community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/BatchingPhysicalTransactionAppender.java

```diff
@@ -137,37 +81,93 @@ protected void emptyBufferIntoChannel() throws IOException
     {
+        // There's a benign race here, where we add our link before we update our next pointer.
+        // This is okay, however, because unparkAll() spins when it sees a null next pointer.
         ThreadLink threadLink = new ThreadLink( Thread.currentThread() );
         threadLink.next = threadLinkHead.getAndSet( threadLink );
+        int waitTicks = 127;
+
+        do
+        {
+            if ( forceLock.tryLock() )
+            {
+                try
+                {
+                    forceLog();
+                }
+                finally
+                {
+                    forceLock.unlock();
+
+                    // We've released the lock, so unpark anyone who might have decided park while we were working.
+                    // The most recently parked thread is the one most likely to still have warm caches, so that's
+                    // the one we would prefer to unpark. Luckily, the stack nature of the ThreadLinks makes it easy
+                    // to get to.
+                    ThreadLink nextWaiter = threadLinkHead.get();
+                    nextWaiter.unpark();
+                }
+            }
+            else
+            {
+                waitTicks = waitForLogForce( waitTicks );
+            }
+        }
+        while ( !threadLink.done );
+    }
+
+    private void forceLog() throws IOException
+    {
+        ThreadLink links = threadLinkHead.getAndSet( ThreadLink.END );
+
+        force();
+
+        unparkAll( links );
+    }
+
+    private void unparkAll( ThreadLink links )
+    {
+        do
+        {
+            links.done = true;
+            links.unpark();
+            ThreadLink tmp;
+            do
+            {
+                // Spin on this because of the racy update when consing.
+                tmp = links.next;
+            }
+            while ( tmp == null );
+            links = tmp;
+        }
+        while ( links != ThreadLink.END );
+    }
 
-        // Stay a while and listen... while:
-        while (  // the forcer hasn't yet caught up with me
-                 (ticket > forceTicket.get() ||
-                 // OR I've wrapped around Long.MAX_VALUE
-                 !haveSameSign( ticket, forceTicket.get() )) &&
+    private int waitForLogForce( int waitTicks )
+    {
+        waitTicks &= 127;
 
-                 // AND this appender hasn't yet been shut down
-                 !shutDown &&
-                 // AND the forcer is of good health
-                 forceThread.checkHealth() )
+        // We do this fancy spin to create CPU pipeline stalls in which other threads can run a few instructions.
+        // The hope is that those other threads might make enough progress to either finish our work, or allow us
+        // to continue.
+        if ( ThreadLocalRandom.current().nextBoolean() )
+        {
+            return waitTicks - 1;
+        }
+        if ( waitTicks < 4 && waitTicks != 0 )
+        {
+            Thread.yield();
```

Comment:
We could experiment with just going straight to park(), instead of spending these yield() calls.